scripts/fbi/hate_crime/download_publication_data.py

```diff
@@ -0,0 +1,363 @@
+import requests
+import os
+from bs4 import BeautifulSoup
+import time
+import json
+from absl import app
+from absl import flags
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer('start_year',
+                     2005,
+                     'Start year(Min. 2005)',
+                     lower_bound=2005)
+flags.DEFINE_integer('end_year', 2019, 'End year(Min. 2005)', lower_bound=2005)
+flags.DEFINE_boolean('force_fetch', True,
+                     'forces api query and not return cached result')
+flags.DEFINE_string('store_path', './tmp', 'Path to store the output')
+flags.DEFINE_string('cache_path', './tmp/.cache_html/',
+                    'Path to store the cached html files')
+flags.DEFINE_string('base_url', 'https://ucr.fbi.gov/hate-crime',
+                    'The base url containing list of URLs for each year')
+
+# TODO(issue id here) store http errors
+# TODO(issue id here) log no access URL found
+
+# Headers to make request
+_HEADERS = {
+    'User-Agent':
+        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.55 Safari/537.36'
+}
+"""
+The file names of excel files changes across the years.
+The URLs to access table appear in sequence of table number,
+hence code uses enumerate to assign the table number.
+years_with_no_table0 list help make this correction.
+"""
+_YEARS_WITH_NO_TABLE0 = ('2004', '2005', '2006', '2007')
+
+# dictionary to access status across method calls
+# TODO(issue id here) remove this as global variable
+_STATUS_DICT = {
+    'year_list': [str(x) for x in range(2005, 2020)],
+    'output_dir': './source_data',
+    'cache_dir': './source_data/.cache_html/',
+    'force_fetch': True,
+    'base_url': 'https://ucr.fbi.gov/hate-crime',
+    'session': None,
+    'cur_year': '',
+    'cur_year_url': '',
+    'cur_access_url': '',
+    'table_html_url_list': [],
+    'download_urls': {}
+}
+
+
+def _get_abs_url(base_url: str, cur_url: str) -> str:
+    """Convert relative URL to absolute URL.
+  """
+    ret_url = cur_url
+    if not cur_url.startswith('http'):
+        if cur_url.startswith('/'):
+            ret_url = requests.compat.urljoin(base_url, cur_url)
+        else:
+            ret_url = base_url + '/' + cur_url
+    return ret_url
+
+
+def request_html_soup(cur_session, url: str) -> tuple:
+    """Get html text and bs4 object of the html from URL.
+    Args:
+        cur_session: requests session object which should be used to make url call.
+        url: The URL to be fetched.
+    Returns:
+        A tuple:
+            resp_html: The response HTML page text version.
+            resp_soup: BeautifulSoup object of the response HTML.
+  """
+    req = cur_session.get(url)
+    if req.status_code == requests.codes.ok:
+        resp_html = req.text
+        resp_soup = BeautifulSoup(resp_html, features="html.parser")
+    else:
+        resp_html = None
+        resp_soup = None
+
+        print('HTTP status code: ' + str(req.status_code))
```

Comment:
Can remove empty line 100 and add an ERROR prefix to the print mesage with the url attempted.

Reply:
fixed.